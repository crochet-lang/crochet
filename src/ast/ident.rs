use crate::ast::span::Span;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Ident {
    pub span: Span,
    pub name: String,
}
