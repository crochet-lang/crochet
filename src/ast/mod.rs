pub mod literal;
pub mod syntax;
