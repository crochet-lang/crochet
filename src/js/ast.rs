use super::super::literal::Lit;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    Decl {
        pattern: Pattern,
        value: Expression,
    },
    Expression {
        expr: Expression,
    },
    Return {
        arg: Expression,   
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expression {
    Call {
        func: Box<Expression>,
        args: Vec<Expression>,
    },
    Function {
        params: Vec<Param>,
        body: Vec<Statement>,
        r#async: bool,
    },
    Ident {
        name: String,
    },
    Literal(Lit),
    Binary {
        op: BinaryOp,
        left: Box<Expression>,
        right: Box<Expression>,
    },
    Unary {
        op: UnaryOp,
        arg: Box<Expression>,
    },
    IfElse {
        cond: Box<Expression>,
        consequent: Vec<Statement>,
        alternate: Vec<Statement>,
    },
    Object {
        properties: Vec<Property>,
    },
    Await {
        expr: Box<Expression>,
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Pattern {
    Ident { name: String },
    // TODO: add more patterns later
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Param {
    Ident { name: String },
    Rest { name: String },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Property {
    pub name: String,
    pub value: Expression,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Exp,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum UnaryOp {
    Neg,
    Not,
}
