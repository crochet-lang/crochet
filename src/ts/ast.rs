use itertools::{join, Itertools};
use std::fmt;

use super::super::literal::Literal;
use super::super::types::Primitive;

pub struct TsQualifiedType {
    pub ty: TsType,
    pub type_params: Vec<i32>,
}

#[derive(Debug)]
pub enum TsType {
    Prim(Primitive),
    Var(String),
    Lit(Literal),
    Func {
        params: Vec<Param>,
        ret: Box<TsType>,
    },
    Union(Vec<TsType>),
    Obj(Vec<TsObjProp>),
    Alias {
        name: String,
        type_params: Vec<TsType>,
    },
}

impl fmt::Display for TsQualifiedType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let chars: Vec<_> = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
            .chars()
            .collect();

        if self.type_params.len() > 0 {
            let type_params = self
                .type_params
                .iter()
                .map(|id| {
                    let id = chars.get(id.to_owned() as usize).unwrap();
                    format!("{id}")
                })
                .join(", ");
            write!(f, "<{type_params}>{}", self.ty)
        } else {
            write!(f, "{}", self.ty)
        }
    }
}

impl fmt::Display for TsType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            TsType::Var(name) => write!(f, "{}", name),
            TsType::Prim(prim) => write!(f, "{}", prim),
            TsType::Lit(lit) => write!(f, "{}", lit),
            TsType::Func { params, ret } => {
                let params = params
                    .iter()
                    // TODO: use write! to format the params more directly instead of
                    // using the intermediary format!
                    .map(|Param { name, ty }| format!("{name}: {ty}").to_owned())
                    .join(", ");
                write!(f, "({params}) => {ret}")
            }
            TsType::Union(types) => write!(f, "{}", join(types, " | ")),
            TsType::Obj(props) => {
                let props = props
                    .iter()
                    .map(|TsObjProp { name, ty }| format!("{name}: {ty}"))
                    .join(", ");

                // TODO: output multi-line object types
                write!(f, "{{{props}}}")
            }
            TsType::Alias {name, type_params} => {
                let type_params = type_params.iter().join(", ");
                write!(f, "{name}<{type_params}>")
            }
        }
    }
}

#[derive(Debug)]
pub struct Param {
    pub name: String,
    pub ty: TsType,
}

#[derive(Debug)]
pub struct TsObjProp {
    pub name: String,
    pub ty: TsType,
}
