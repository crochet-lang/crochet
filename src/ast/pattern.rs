use crate::ast::span::Span;
use crate::ast::ident::Ident;
use crate::ast::types::TypeAnn;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Pattern {
    Ident(BindingIdent),
    Rest(RestPat),

    // Array(ArrayPat),
    // Object(ObjectPat),
    // Assign(AssignPat),
}

impl Pattern {
    pub fn span(&self) -> Span {
        match self {
            Pattern::Ident(ident) => ident.span.to_owned(),
            Pattern::Rest(rest) => rest.span.to_owned(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BindingIdent {
    pub span: Span,
    pub id: Ident,
    pub type_ann: Option<TypeAnn>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RestPat {
    pub span: Span,
    pub arg: Box<Pattern>,
    pub type_ann: Option<TypeAnn>,
}
