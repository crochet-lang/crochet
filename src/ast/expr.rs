use crate::ast::ident::Ident;
use crate::ast::jsx::JSXElement;
use crate::ast::literal::Lit;
use crate::ast::pattern::Pattern;
use crate::ast::span::Span;
use crate::ast::types::{TypeAnn, TypeParam};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    VarDecl {
        span: Span,
        pattern: Pattern,
        init: Option<Expr>,
        declare: bool,
    },
    TypeDecl {
        span: Span,
        declare: bool,
        id: Ident,
        type_ann: TypeAnn,    
        type_params: Option<Vec<TypeParam>>,    
    },
    Expr {
        span: Span,
        expr: Expr,
    }, // NOTE: does not include Expr::Let
}

// #[derive(Debug, Clone, PartialEq, Eq)]
// struct Block {
//     pub span: Span,
//     pub stmts: Vec<Expr>,
// }

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct App {
    pub span: Span,
    pub lam: Box<Expr>,
    pub args: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Fix {
    pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IfElse {
    pub span: Span,
    pub cond: Box<Expr>,
    pub consequent: Box<Expr>,
    pub alternate: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Lambda {
    pub span: Span,
    pub params: Vec<Pattern>,
    pub body: Box<Expr>,
    pub is_async: bool,
    pub return_type: Option<TypeAnn>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Let {
    pub span: Span,
    pub pattern: Option<Pattern>,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Op {
    pub span: Span,
    pub op: BinOp,
    pub left: Box<Expr>,
    pub right: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Obj {
    pub span: Span,
    pub props: Vec<Property>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Await {
    pub span: Span,
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Tuple {
    pub span: Span,
    pub elems: Vec<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Member {
    pub span: Span,
    pub obj: Box<Expr>,
    pub prop: MemberProp,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MemberProp {
    Ident(Ident),
    Computed(ComputedPropName),
}

impl MemberProp {
    pub fn name(&self) -> String {
        match self {
            MemberProp::Ident(Ident { name, .. }) => name.to_owned(),
            MemberProp::Computed(_) => todo!(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ComputedPropName {
    pub span: Span, // includes enclosing []
    pub expr: Box<Expr>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    App(App),
    Fix(Fix),
    Ident(Ident),
    IfElse(IfElse),
    JSXElement(JSXElement),
    Lambda(Lambda),
    Let(Let),
    Lit(Lit),
    Op(Op),
    Obj(Obj),
    Await(Await),
    Tuple(Tuple),
    Member(Member),
}

impl Expr {
    pub fn span(&self) -> Span {
        match &self {
            Expr::App(app) => app.span.to_owned(),
            Expr::Fix(fix) => fix.span.to_owned(),
            Expr::Ident(ident) => ident.span.to_owned(),
            Expr::IfElse(if_else) => if_else.span.to_owned(),
            Expr::JSXElement(elem) => elem.span.to_owned(),
            Expr::Lambda(lam) => lam.span.to_owned(),
            Expr::Let(r#let) => r#let.span.to_owned(),
            Expr::Lit(lit) => lit.span(),
            Expr::Op(op) => op.span.to_owned(),
            Expr::Obj(obj) => obj.span.to_owned(),
            Expr::Await(r#await) => r#await.span.to_owned(),
            Expr::Tuple(tuple) => tuple.span.to_owned(),
            Expr::Member(member) => member.span.to_owned(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    EqEq,
    NotEq,
    Gt,
    GtEq,
    Lt,
    LtEq,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Property {
    pub span: Span,
    pub name: String,
    pub value: Expr,
}
