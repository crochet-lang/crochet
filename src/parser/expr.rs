use chumsky::prelude::*;

use crate::ast::*;
use crate::parser::jsx::jsx_parser;
use crate::parser::pattern::pattern_parser;
use crate::parser::type_params::type_params;
use crate::parser::types::type_parser;
use crate::parser::util::just_with_padding;

pub fn expr_parser() -> BoxedParser<'static, char, Expr, Simple<char>> {
    let type_ann = type_parser();
    let pattern = pattern_parser();

    let ident = text::ident().map_with_span(|name, span| Ident { span, name });

    let r#true =
        just_with_padding("true").map_with_span(|_, span| Expr::Lit(Lit::bool(true, span)));
    let r#false =
        just_with_padding("false").map_with_span(|_, span| Expr::Lit(Lit::bool(false, span)));
    let r#bool = choice((r#true, r#false));

    let int = text::int::<char, Simple<char>>(10)
        .map_with_span(|value, span| Expr::Lit(Lit::num(value, span)));
    let real = text::int(10)
        .chain(just('.'))
        .chain::<char, _, _>(text::digits(10))
        .collect::<String>()
        .map_with_span(|value, span| Expr::Lit(Lit::num(value, span)));
    let num = choice((real, int));

    let r#str = just("\"")
        .ignore_then(filter(|c| *c != '"').repeated().at_least(1))
        .then_ignore(just("\""))
        .collect::<String>()
        .map_with_span(|value, span| Expr::Lit(Lit::str(value, span)));

    let parser = recursive(|expr: Recursive<'_, char, Expr, Simple<char>>| {
        // TODO: support recursive functions to be declared within another function
        // let let_rec = ...

        let block = just("let")
            .ignore_then(pattern.clone())
            .then_ignore(just_with_padding("="))
            .or_not()
            .then(expr.clone())
            .separated_by(just_with_padding(";"))
            .then(just_with_padding(";").or_not())
            .delimited_by(just_with_padding("{"), just_with_padding("}"))
            .map(|(lets, trailing_semi)| {
                let mut iter = lets.iter().rev();

                // TODO: if `lets` is empty then we should return the empty type
                
                let last = match trailing_semi {
                    Some(_) => {
                        Expr::Empty(Empty {
                            span: 0..0,
                        })
                    }
                    None => {
                        match iter.next() {
                            Some(term) => match term {
                                // TODO: if we do get a `let` last, we should be able to type
                                // is as `empty`
                                (Some(_), _) => panic!("Didn't expect `let` here"),
                                (_, expr) => expr.clone(),
                            },
                            None => {
                                Expr::Empty(Empty {
                                    span: 0..0,
                                })
                            },
                        }
                    }
                };

                let result: Expr = iter.fold(last, |body, (pattern, value)| {
                    let start = match pattern {
                        Some(pattern) => pattern.span().start,
                        None => value.span().start,
                    };
                    let end = body.span().end;

                    Expr::Let(Let {
                        span: start..end,
                        pattern: pattern.to_owned(),
                        init: Box::new(value.to_owned()),
                        body: Box::new(body),
                    })
                });

                result
            });

        let if_else = recursive(|if_else| {
            just_with_padding("if")
                .ignore_then(expr.clone())
                .then(block.clone())
                .then(just_with_padding("else").ignore_then(block.clone().or(if_else)).or_not())
                .map_with_span(|((cond, cons), alt), span: Span| {
                    Expr::IfElse(IfElse {
                        span,
                        cond: Box::from(cond),
                        consequent: Box::from(cons),
                        alternate: alt.map(Box::from)
                    })
                })
        });

        let key_value_prop = text::ident()
            .then_ignore(just_with_padding(":"))
            .then(expr.clone())
            .map_with_span(|(name, value), span: Span| {
                Prop::KeyValue(KeyValueProp { span, name, value })
            });

        let shorthand_prop =
            text::ident().map_with_span(|name, span: Span| Prop::Shorthand(Ident { span, name }));

        let prop = choice((key_value_prop, shorthand_prop));

        let obj = prop
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("{"), just_with_padding("}"))
            .map_with_span(|props, span: Span| Expr::Obj(Obj { span, props }));

        let tuple = expr
            .clone()
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("["), just_with_padding("]"))
            .map_with_span(|elems, span: Span| Expr::Tuple(Tuple { span, elems }));

        let atom = choice((
            if_else,
            r#bool,
            num,
            r#str,
            ident.map(Expr::Ident),
            obj,
            tuple,
            jsx_parser(expr.clone().boxed()),
            expr.clone()
                .delimited_by(just_with_padding("("), just_with_padding(")")),
        ));

        let mem = atom
            .clone()
            .then(
                just_with_padding(".")
                    .ignore_then(text::ident())
                    .map_with_span(|name, span: Span| {
                        let prop = MemberProp::Ident(Ident {
                            span: span.clone(),
                            name,
                        });
                        (prop, span)
                    })
                    .repeated(),
            )
            .foldl(|f, (prop, span)| {
                let start = f.span().start;
                let end = span.end;
                let span = start..end;

                Expr::Member(Member {
                    span,
                    obj: Box::new(f),
                    prop,
                })
            });

        let app = mem
            .clone()
            .then(
                expr.clone()
                    .separated_by(just_with_padding(","))
                    .allow_trailing()
                    .delimited_by(just_with_padding("("), just_with_padding(")"))
                    .map_with_span(|args, span: Span| (args, span))
                    .repeated(),
            )
            .foldl(|f, (args, span)| {
                let start = f.span().start;
                let end = span.end;
                let span = start..end;

                Expr::App(App {
                    span,
                    lam: Box::new(f),
                    args,
                })
            });

        // Application is higher precedence than `await`
        let r#await = just_with_padding("await")
            .or_not()
            .then(app.clone())
            .map_with_span(|(option, arg), span: Span| match option {
                Some(_) => Expr::Await(Await {
                    span,
                    expr: Box::from(arg),
                }),
                None => arg,
            });

        let product = r#await
            .clone()
            .then(
                choice((
                    just_with_padding("*").to(BinOp::Mul),
                    just_with_padding("/").to(BinOp::Div),
                ))
                .then(r#await.clone())
                .repeated(),
            )
            .foldl(|left, (op, right)| {
                // atoms are already using source spans since they're WithSpan<Expr>
                let span = left.span().start..right.span().end;
                Expr::Op(Op {
                    span,
                    op,
                    left: Box::from(left),
                    right: Box::from(right),
                })
            });

        let sum = product
            .clone()
            .then(
                choice((
                    just_with_padding("+").to(BinOp::Add),
                    just_with_padding("-").to(BinOp::Sub),
                ))
                .then(product.clone())
                .repeated(),
            )
            .foldl(|left, (op, right)| {
                // products are already using source spans since they're WithSpan<Expr>
                let span = left.span().start..right.span().end;
                Expr::Op(Op {
                    span,
                    op,
                    left: Box::from(left),
                    right: Box::from(right),
                })
            });

        // TODO: logic operators
        let comp = sum
            .clone()
            .then(
                choice((
                    just_with_padding("==").to(BinOp::EqEq),
                    just_with_padding("!=").to(BinOp::NotEq),
                    // Must appear before ">"
                    just_with_padding(">=").to(BinOp::GtEq),
                    just_with_padding(">").to(BinOp::Gt),
                    // Must appear before "<"
                    just_with_padding("<=").to(BinOp::LtEq),
                    just_with_padding("<").to(BinOp::Lt),
                ))
                .then(sum.clone())
                .repeated(),
            )
            .foldl(|left, (op, right)| {
                // sums are already using source spans since they're WithSpan<Expr>
                let span = left.span().start..right.span().end;
                Expr::Op(Op {
                    span,
                    op,
                    left: Box::from(left),
                    right: Box::from(right),
                })
            });

        let param_list = pattern_parser()
            .separated_by(just_with_padding(","))
            .allow_trailing()
            .delimited_by(just_with_padding("("), just_with_padding(")"));

        let lam = just_with_padding("async")
            .or_not()
            .then(type_params(type_ann.clone().boxed()).or_not())
            .then(param_list)
            .then(just_with_padding(":").ignore_then(type_ann).or_not())
            .then_ignore(just_with_padding("=>"))
            .then(choice((block.clone(), expr.clone())))
            .map_with_span(
                |((((is_async, type_params), args), return_type), body), span: Span| {
                    Expr::Lambda(Lambda {
                        span,
                        params: args,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                        return_type,
                        type_params,
                    })
                },
            );

        choice((lam, block, comp))
    });

    parser.boxed()
}
