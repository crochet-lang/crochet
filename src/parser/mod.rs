pub mod expr;
pub mod jsx;
pub mod pattern;
pub mod types;

use expr::expr_parser;
use pattern::pattern_parser;

use chumsky::prelude::*;
use chumsky::primitive::*;
use chumsky::text::Padded;

use crate::ast::*;

pub fn just_with_padding(inputs: &str) -> Padded<Just<char, &str, Simple<char>>> {
    just(inputs).padded()
}

pub fn parser() -> impl Parser<char, Program, Error = Simple<char>> {
    // We use `just` instead of `just_with_padding` here to ensure that
    // the span doesn't include leading whitespace.
    // TODO: add `declare` syntax
    let decl = just("let")
        .ignore_then(just_with_padding("rec").or_not())
        .then(pattern_parser())
        .then_ignore(just_with_padding("="))
        .then(expr_parser())
        .map_with_span(|((rec, pattern), value), span: Span| -> Statement {
            match rec {
                Some(_) => {
                    Statement::Decl {
                        span,
                        pattern: pattern.clone(),
                        // `let fib = fix((fib) => (n) => ...)`
                        // TODO: Fix always wraps a lambda
                        value: Expr::Fix(Fix {
                            span: value.span(),
                            expr: Box::from(Expr::Lambda(Lambda {
                                span: value.span(),
                                args: vec![pattern.clone()],
                                body: Box::from(value),
                                is_async: false,
                            })),
                        }),
                    }
                }
                None => Statement::Decl {
                    span,
                    pattern,
                    value,
                },
            }
        });

    let program = choice((
        decl,
        expr_parser().map_with_span(|expr, span: Span| Statement::Expr { expr, span }),
    ))
    .padded()
    .repeated()
    .map(|body| Program { body });

    program.then_ignore(end())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(input: &str) -> Program {
        parser().parse(input).unwrap()
    }

    #[test]
    fn literals() {
        insta::assert_debug_snapshot!(parse("10"));
        insta::assert_debug_snapshot!(parse("1.23"));
        insta::assert_debug_snapshot!(parse("\"hello\""));
    }

    #[test]
    fn operations() {
        insta::assert_debug_snapshot!(parse("1 + 2 - 3"));
        insta::assert_debug_snapshot!(parse("x * y / z"));
        insta::assert_debug_snapshot!(parse("(a + b) * c"));
        insta::assert_debug_snapshot!(parse("a == b"));
        insta::assert_debug_snapshot!(parse("a != b"));
        insta::assert_debug_snapshot!(parse("a > b"));
        insta::assert_debug_snapshot!(parse("a >= b"));
        insta::assert_debug_snapshot!(parse("a < b"));
        insta::assert_debug_snapshot!(parse("a <= b"));
        insta::assert_debug_snapshot!(parse("let cond = a != b"));
    }

    #[test]
    fn function_definition() {
        insta::assert_debug_snapshot!(parse("(a, b) => c"));
        insta::assert_debug_snapshot!(parse("() => 10"));
        insta::assert_debug_snapshot!(parse("(a) => \"hello\""));
    }

    #[test]
    fn async_await() {
        insta::assert_debug_snapshot!(parse("async () => 10"));
        insta::assert_debug_snapshot!(parse("let foo = async () => { await 10 }"));
        insta::assert_debug_snapshot!(parse("let foo = async () => await a + await b"));
        insta::assert_debug_snapshot!(parse("let foo = async () => await bar()"));
    }

    #[test]
    fn function_application() {
        insta::assert_debug_snapshot!(parse("foo()"));
        insta::assert_debug_snapshot!(parse("foo(a, b)"));
        insta::assert_debug_snapshot!(parse("foo(10, \"hello\")"));
        insta::assert_debug_snapshot!(parse("f(x)(g(x))"));
    }

    #[test]
    fn declarations() {
        insta::assert_debug_snapshot!(parse("let x = 5"));
        insta::assert_debug_snapshot!(parse("let x = (a, b) => a + b"));
        insta::assert_debug_snapshot!(parse("let foo = let x = 5 in x"));
        insta::assert_debug_snapshot!(parse("let rec f = () => f()")); // recursive
    }

    #[test]
    fn top_level_expressions() {
        insta::assert_debug_snapshot!(parse("a + b"));
        insta::assert_debug_snapshot!(parse("123\n\"hello\""));
    }

    #[test]
    fn if_else() {
        insta::assert_debug_snapshot!(parse("if (true) { 5 } else { 10 }"));
    }

    #[test]
    fn objects() {
        insta::assert_debug_snapshot!(parse("{x: 5, y: 10}"));
    }

    #[test]
    fn jsx() {
        insta::assert_debug_snapshot!(parse("<Foo>Hello</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>{bar}</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>Hello {world}!</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>{<Bar>{baz}</Bar>}</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo></Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo bar={baz} />"));
        insta::assert_debug_snapshot!(parse("<Foo msg=\"hello\" bar={baz}></Foo>"));
    }

    #[test]
    fn type_annotations() {
        insta::assert_debug_snapshot!(parse("let x: number = 5"));
        insta::assert_debug_snapshot!(parse("let msg: string = \"hello\""));
        insta::assert_debug_snapshot!(parse("let add = (a: number, b: number) => a + b"));
        insta::assert_debug_snapshot!(parse("let p: Point = {x: 5, y: 10}"));
        insta::assert_debug_snapshot!(parse("let FOO: \"foo\" = \"foo\""));
    }
}
