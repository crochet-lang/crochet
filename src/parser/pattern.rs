use chumsky::prelude::*;

use crate::ast::*;
use crate::parser::types::*;
use crate::parser::util::just_with_padding;

// NOTE: Destructuring assignments admits different patterns from destructuring
// function params.  We'll need to have different parsers for those.
pub fn pattern_parser() -> BoxedParser<'static, char, Pattern, Simple<char>> {
    let type_ann = type_parser();
    let mut top_level = true;

    let parser = recursive(|pat| {
        let ident_pat = text::ident()
            .map_with_span(|name, span| Ident { name, span })
            .then(just_with_padding(":").ignore_then(type_ann).or_not())
            .map_with_span(|(id, type_ann), span: Span| {
                Pattern::Ident(BindingIdent { span, id, type_ann })
            })
            .padded();

        // NOTE: rest patterns must appear instead of other patterns
        // It's an error for them appear at the top-level, e.g.
        // let ...x = [1,2,3];
        // Also, type annotation can only appear at the top-level as well.
        let rest_pat = just("...")
            .ignore_then(pat.clone())
            .map_with_span(|arg, span| RestPat {
                span,
                arg: Box::from(arg),
                type_ann: None,
            });

        let array_pat = pat.clone()
            .separated_by(just_with_padding(","))
            .delimited_by(just_with_padding("["), just_with_padding("]"))
            .map_with_span(|elems, span| {
                Pattern::Array(ArrayPat {
                    span,
                    // The reason why each elem is wrapped in Some() is that
                    // it's possible to have a pattern with skipped elements
                    // althought the parser doesn't support this yet.
                    elems: elems.iter().cloned().map(Some).collect(),
                    optional: false,
                    type_ann: None,
                })
            });

        let key_value_pat_prop = text::ident()
            .map_with_span(|name, span| Ident { span, name })
            .then_ignore(just_with_padding(":"))
            .then(pat.clone())
            .map(|(key, value)| {
                ObjectPatProp::KeyValue(KeyValuePatProp {
                    key,
                    value: Box::from(value),
                })
            });

        // TODO: support default values
        let assign_pat_prop = text::ident()
            .map_with_span(|name, span| Ident { span, name })
            .map_with_span(|key, span| {
                ObjectPatProp::Assign(AssignPatProp {
                    span,
                    key,
                    value: None,
                })
            });

        // NOTE: There can only be a single rest element and it must be last
        let obj_pat_prop = choice((
            rest_pat.clone().map(ObjectPatProp::Rest),
            key_value_pat_prop,
            assign_pat_prop,
        ))
        .separated_by(just_with_padding(","))
        .delimited_by(just_with_padding("{"), just_with_padding("}"))
        .map_with_span(|props, span| {
            Pattern::Object(ObjectPat {
                span,
                props,
                optional: false,
                type_ann: None,
            })
        });

        top_level = false;

        choice((ident_pat, array_pat, obj_pat_prop, rest_pat.map(Pattern::Rest)))
    });

    parser.boxed()
}
