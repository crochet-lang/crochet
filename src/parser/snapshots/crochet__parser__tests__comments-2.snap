---
source: src/parser/mod.rs
expression: "parse(r#\"\n        // x is 5\n        // y is 10\n        let p = {x: 5, y: 10}\n        \"#)"
---
Program {
    body: [
        VarDecl {
            span: 9..76,
            comments: Some(
                [
                    Comment {
                        span: 9..27,
                        text: " x is 5",
                    },
                    Comment {
                        span: 27..46,
                        text: " y is 10",
                    },
                ],
            ),
            pattern: Ident(
                BindingIdent {
                    span: 50..51,
                    id: Ident {
                        span: 50..51,
                        name: "p",
                    },
                    type_ann: None,
                },
            ),
            init: Some(
                Obj(
                    Obj {
                        span: 54..76,
                        props: [
                            KeyValue(
                                KeyValueProp {
                                    span: 55..59,
                                    name: "x",
                                    value: Lit(
                                        Num(
                                            Num {
                                                span: 58..59,
                                                value: "5",
                                            },
                                        ),
                                    ),
                                },
                            ),
                            KeyValue(
                                KeyValueProp {
                                    span: 61..66,
                                    name: "y",
                                    value: Lit(
                                        Num(
                                            Num {
                                                span: 64..66,
                                                value: "10",
                                            },
                                        ),
                                    ),
                                },
                            ),
                        ],
                    },
                ),
            ),
            declare: false,
        },
    ],
}
